### DISCUSSION TITLE
Topic data pagination usability

### DISCUSSION THREAD

**Discussion by dweber019** [tchiotludo/akhq#1507] [Q&A] [answered]
IMHO the pagination usability of the topic data view is broken.

I have a topic with 45 events
![CleanShot 2023-06-14 at 07 59 49](https://github.com/tchiotludo/akhq/assets/1021324/3b3316a5-d84d-401d-8b33-f8ebb5398975)

But if I navigate to the topic data view I only see 2 events right away
![CleanShot 2023-06-14 at 08 00 29](https://github.com/tchiotludo/akhq/assets/1021324/bdcaa34f-0686-4221-9d24-4e65f321b6b6)

I can see on the pager that there should be 45 in the right upper corner.

When I navigate to the next page I can still see only 2 events, these are two other events
![CleanShot 2023-06-14 at 08 02 08](https://github.com/tchiotludo/akhq/assets/1021324/c5b223ee-3219-498c-8a36-91cd1ff5b5cc)

I would expect to see akhq.topic-data.size=50 events according to https://akhq.io/docs/configuration/akhq.html#topic-data per page.

Expectation:
1. akhq.topic-data.size works, if available 50 events per page are displayed
2. Pagination has a not only a next page button but a previous button too

Addition:
- I think it would even make sense to have the same paging experience as on the topic list view.
- I just realised that if I configure akhq.ui-options.topic-data.sort=OLDEST the akhq.topic-data.size is at least respected but never the less as the default is NEWEST, it's no a confusing experience

Context:
From time to time I give some Kafka courses and this usability is always an issue for the participants.

**Comment by tchiotludo** [comment, marked as answer]
Hey, 

It's totally impossible due to Kafka Api: 
- Kafka give us a number of record based on start offset and end offset (what is displayed on pagination). 
- But these number are only bound number of offset and there no way to know what happen inside these bounds.
- Kafka have internal process that could leave empty offset that you will never be able to pull 
  - each transaction will add empty offset for the BEGIN and the END at least 
  - each compacted topic will remove duplicate key (what seems to be your case on the screenshot with multiple `null` key)
- it's why I added a `≈` everywhere, Kafka don't allow any correct pagination

**Reply by dweber019** [reply]
The topic is not compacted.
<img width="1316" alt="CleanShot 2023-06-21 at 20 02 38@2x" src="https://github.com/tchiotludo/akhq/assets/1021324/c30019e2-6b24-4d68-a8fb-93398531c764">

Are there any other reasons for this implementation / limitation?
I'm asking because other tools are able to do this like Kafka UI and wondering if there is a good reason why not to change this. Maybe there is something I'm not aware of 😅

**Reply by tchiotludo** [reply]
in that case, the number of records should be nearly relevant I think, but still not exact (depending on your workload, for example, transactions, ...)

**Reply by dweber019** [reply]
Sorry maybe my request was a bit misleading. The number next to the paging button isn't the issue for me. An issue is that only 2 messages are displayed per page and that we aren't able to page back to the previous page.
